File: services/collectors/official_marvel_rivals/news_fetcher.py
from __future__ import annotations

import logging
from dataclasses import dataclass
from urllib.parse import urldefrag, urljoin, urlparse

import httpx
from bs4 import BeautifulSoup, Tag


logger = logging.getLogger(__name__)

USER_AGENT = (
    "Mozilla/5.0 (compatible; MarvelRivalsUACollector/1.0; "
    "+https://www.marvelrivals.com/news/)"
)
REQUEST_TIMEOUT_SECONDS = 20.0


@dataclass(frozen=True)
class NewsArticleSummary:
    title: str
    canonical_url: str
    article_url: str
    raw_date: str | None
    raw_excerpt: str | None
    cover_image_url: str | None


async def fetch_html(url: str) -> str | None:
    try:
        async with httpx.AsyncClient(
            follow_redirects=True,
            timeout=httpx.Timeout(REQUEST_TIMEOUT_SECONDS),
            headers={"User-Agent": USER_AGENT},
        ) as client:
            response = await client.get(url)
            response.raise_for_status()
            return response.text
    except httpx.HTTPError as exc:
        logger.warning("Failed to fetch %s: %s", url, exc)
        return None


class OfficialNewsFetcher:
    def __init__(self, news_url: str) -> None:
        self.news_url = news_url

    async def fetch_recent_articles(self) -> list[NewsArticleSummary]:
        html = await fetch_html(self.news_url)
        if html is None:
            return []

        try:
            articles = parse_official_news_list(html, self.news_url)
        except Exception:
            logger.exception("Failed to parse official Marvel Rivals news list")
            return []

        if not articles:
            logger.warning("No articles found on official news page %s", self.news_url)
        else:
            logger.info("Found %s official Marvel Rivals news articles", len(articles))

        return articles


def parse_official_news_list(html: str, base_url: str) -> list[NewsArticleSummary]:
    soup = BeautifulSoup(html, "html.parser")
    articles: list[NewsArticleSummary] = []
    seen_urls: set[str] = set()

    for item in soup.select("a.list-item[href]"):
        article = _summary_from_card(item, base_url)
        if article is None or article.canonical_url in seen_urls:
            continue

        seen_urls.add(article.canonical_url)
        articles.append(article)

    if articles:
        return articles

    logger.warning("Official news list cards were not found. Falling back to link scanning.")
    for link in soup.select("a[href]"):
        article = _summary_from_link(link, base_url)
        if article is None or article.canonical_url in seen_urls:
            continue

        seen_urls.add(article.canonical_url)
        articles.append(article)

    return articles


def _summary_from_card(item: Tag, base_url: str) -> NewsArticleSummary | None:
    href = item.get("href")
    article_url = _normalize_article_url(str(href), base_url) if href else None
    if article_url is None:
        return None

    title = _clean_text(_first_text(item, "h1, h2, h3") or item.get("title") or "")
    if not title:
        title = _title_from_url(article_url)

    raw_excerpt = _clean_text(_first_text(item, "p") or "")
    raw_date = _clean_text(_first_text(item, "time, .date, .time") or "")
    cover_image_url = _normalize_media_url(_first_image_src(item), base_url)

    return NewsArticleSummary(
        title=title,
        canonical_url=article_url,
        article_url=article_url,
        raw_date=raw_date or None,
        raw_excerpt=raw_excerpt or None,
        cover_image_url=cover_image_url,
    )


def _summary_from_link(link: Tag, base_url: str) -> NewsArticleSummary | None:
    href = link.get("href")
    article_url = _normalize_article_url(str(href), base_url) if href else None
    if article_url is None:
        return None

    title = _clean_text(link.get_text(" ", strip=True) or str(link.get("title") or ""))
    if not title:
        title = _title_from_url(article_url)

    return NewsArticleSummary(
        title=title,
        canonical_url=article_url,
        article_url=article_url,
        raw_date=None,
        raw_excerpt=None,
        cover_image_url=_normalize_media_url(_first_image_src(link), base_url),
    )


def _normalize_article_url(value: str, base_url: str) -> str | None:
    absolute_url = urljoin(base_url, value.strip())
    absolute_url, _fragment = urldefrag(absolute_url)
    parsed = urlparse(absolute_url)

    if parsed.scheme not in {"http", "https"}:
        return None
    if parsed.netloc.lower() != "www.marvelrivals.com":
        return None
    if not parsed.path.endswith(".html"):
        return None

    return absolute_url


def _normalize_media_url(value: str | None, base_url: str) -> str | None:
    if not value:
        return None

    absolute_url = urljoin(base_url, value.strip())
    parsed = urlparse(absolute_url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        return None

    return absolute_url


def _first_text(item: Tag, selector: str) -> str | None:
    match = item.select_one(selector)
    return match.get_text(" ", strip=True) if match else None


def _first_image_src(item: Tag) -> str | None:
    image = item.select_one("img[src], img[data-src], img[data-original]")
    if image is None:
        return None

    value = image.get("src") or image.get("data-src") or image.get("data-original")
    return str(value).strip() if value else None


def _clean_text(value: object) -> str:
    return " ".join(str(value).split())


def _title_from_url(url: str) -> str:
    path = urlparse(url).path
    return path.rsplit("/", 1)[-1].replace(".html", "")

File: services/collectors/official_marvel_rivals/news_fetcher.js
import { getText, loadHtml, selectOne } from "../../html.js";
import { getLogger } from "../../logger.js";
import { collapseWhitespace, errorText, rsplitOnce } from "../../pyutils.js";
import { urldefrag, urljoin, urlsplit } from "../../urlutils.js";

const logger = getLogger("services.collectors.official_marvel_rivals.news_fetcher");

export const USER_AGENT =
  "Mozilla/5.0 (compatible; MarvelRivalsUACollector/1.0; +https://www.marvelrivals.com/news/)";
export const REQUEST_TIMEOUT_SECONDS = 20.0;

/**
 * @typedef {{title: string, canonical_url: string, article_url: string,
 *            raw_date: string|null, raw_excerpt: string|null,
 *            cover_image_url: string|null}} NewsArticleSummary
 */

export async function fetchHtml(url) {
  try {
    const response = await fetch(url, {
      redirect: "follow",
      headers: { "User-Agent": USER_AGENT },
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_SECONDS * 1000),
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }
    return await response.text();
  } catch (error) {
    logger.warning(`Failed to fetch ${url}: ${errorText(error)}`);
    return null;
  }
}

export class OfficialNewsFetcher {
  constructor(newsUrl) {
    this.news_url = newsUrl;
  }

  async fetchRecentArticles() {
    const html = await fetchHtml(this.news_url);
    if (html === null) {
      return [];
    }

    let articles;
    try {
      articles = parseOfficialNewsList(html, this.news_url);
    } catch (error) {
      logger.exception("Failed to parse official Marvel Rivals news list", error);
      return [];
    }

    if (!articles.length) {
      logger.warning(`No articles found on official news page ${this.news_url}`);
    } else {
      logger.info(`Found ${articles.length} official Marvel Rivals news articles`);
    }

    return articles;
  }
}

export function parseOfficialNewsList(html, baseUrl) {
  const $ = loadHtml(html);
  const articles = [];
  const seenUrls = new Set();

  $("a.list-item[href]").each((_index, element) => {
    const article = summaryFromCard($, $(element), baseUrl);
    if (article === null || seenUrls.has(article.canonical_url)) {
      return;
    }

    seenUrls.add(article.canonical_url);
    articles.push(article);
  });

  if (articles.length) {
    return articles;
  }

  logger.warning("Official news list cards were not found. Falling back to link scanning.");
  $("a[href]").each((_index, element) => {
    const article = summaryFromLink($, $(element), baseUrl);
    if (article === null || seenUrls.has(article.canonical_url)) {
      return;
    }

    seenUrls.add(article.canonical_url);
    articles.push(article);
  });

  return articles;
}

function summaryFromCard($, item, baseUrl) {
  const href = item.attr("href");
  const articleUrl = href ? normalizeArticleUrl(String(href), baseUrl) : null;
  if (articleUrl === null) {
    return null;
  }

  let title = cleanText(firstText($, item, "h1, h2, h3") ?? item.attr("title") ?? "");
  if (!title) {
    title = titleFromUrl(articleUrl);
  }

  const rawExcerpt = cleanText(firstText($, item, "p") ?? "");
  const rawDate = cleanText(firstText($, item, "time, .date, .time") ?? "");
  const coverImageUrl = normalizeMediaUrl(firstImageSrc($, item), baseUrl);

  return {
    title,
    canonical_url: articleUrl,
    article_url: articleUrl,
    raw_date: rawDate || null,
    raw_excerpt: rawExcerpt || null,
    cover_image_url: coverImageUrl,
  };
}

function summaryFromLink($, link, baseUrl) {
  const href = link.attr("href");
  const articleUrl = href ? normalizeArticleUrl(String(href), baseUrl) : null;
  if (articleUrl === null) {
    return null;
  }

  let title = cleanText(getText(link) || String(link.attr("title") ?? ""));
  if (!title) {
    title = titleFromUrl(articleUrl);
  }

  return {
    title,
    canonical_url: articleUrl,
    article_url: articleUrl,
    raw_date: null,
    raw_excerpt: null,
    cover_image_url: normalizeMediaUrl(firstImageSrc($, link), baseUrl),
  };
}

function normalizeArticleUrl(value, baseUrl) {
  const absoluteUrl = urldefrag(urljoin(baseUrl, value.trim()));
  const parsed = urlsplit(absoluteUrl);

  if (parsed.scheme !== "http" && parsed.scheme !== "https") {
    return null;
  }
  if (parsed.netloc.toLowerCase() !== "www.marvelrivals.com") {
    return null;
  }
  if (!parsed.path.endsWith(".html")) {
    return null;
  }

  return absoluteUrl;
}

function normalizeMediaUrl(value, baseUrl) {
  if (!value) {
    return null;
  }

  const absoluteUrl = urljoin(baseUrl, value.trim());
  const parsed = urlsplit(absoluteUrl);
  if ((parsed.scheme !== "http" && parsed.scheme !== "https") || !parsed.netloc) {
    return null;
  }

  return absoluteUrl;
}

function firstText($, item, selector) {
  const match = selectOne($, selector, item);
  return match ? getText(match) : null;
}

function firstImageSrc($, item) {
  const image = selectOne($, "img[src], img[data-src], img[data-original]", item);
  if (image === null) {
    return null;
  }

  const value = image.attr("src") || image.attr("data-src") || image.attr("data-original");
  return value ? String(value).trim() : null;
}

function cleanText(value) {
  return collapseWhitespace(String(value));
}

function titleFromUrl(url) {
  const path = urlsplit(url).path;
  const last = rsplitOnce(path, "/");
  return (last.length === 2 ? last[1] : last[0]).replaceAll(".html", "");
}
